lib/srv/ctx.go

```diff
@@ -22,121 +22,238 @@ import (
 
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/events"
+	rsession "github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/teleport/lib/sshutils"
 	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
 
 	log "github.com/sirupsen/logrus"
-	"golang.org/x/crypto/ssh"
-	"golang.org/x/crypto/ssh/agent"
 )
 
 var ctxID int32
 
-// subsystemResult is a result of execution of the subsystem
-type subsystemResult struct {
-	err error
+// Server is local or forwarding SSH server.
+type Server interface {
+	// ID is the unique ID of the server.
+	ID() string
+
+	// GetNamespace returns the namespace the server was created in.
+	GetNamespace() string
+
+	// AdvertiseAddr is the publicly addressable address of this server.
+	AdvertiseAddr() string
+
+	// Component is the type of server, forwarding or regular.
+	Component() string
+
+	// PermitUserEnvironment returns if reading environment variables upon
+	// startup is allowed.
+	PermitUserEnvironment() bool
+
+	// EmitAuditEvent an Audit Event from this server.
+	EmitAuditEvent(string, events.EventFields)
+
+	// GetAuditLog returns the Audit Log for this server.
+	GetAuditLog() events.IAuditLog
+
+	// GetAuthService returns an auth.AccessPoint for this server.
+	GetAuthService() auth.AccessPoint
+
+	// GetSessionServer returns a session server.
+	GetSessionServer() rsession.Service
 }
 
-// ctx holds session specific context, such as SSH auth agents
-// PTYs, and other resources. ctx can be used to attach resources
-// that should be closed once the session closes.
-type ctx struct {
+// SessionContext holds session specific context, such as SSH auth agents, PTYs,
+// and other resources. SessionContext also holds a ServerContext which can be
+// used to access resources on the underlying server. SessionContext can also
+// be used to attach resources that should be closed once the session closes.
+type ServerContext struct {
 	*log.Entry
-	// env is a list of environment variables passed to the session
-	env map[string]string
 
-	// srv is a pointer to the server holding the context
-	srv *Server
+	sync.RWMutex
 
-	// server specific incremental session id
-	id int
+	// env is a list of environment variables passed to the session.
+	env map[string]string
 
-	// SSH connection
-	conn *ssh.ServerConn
+	// srv is the server that is holding the context.
+	srv Server
 
-	sync.RWMutex
+	// id is the server specific incremental session id.
+	id int
 
-	// term holds PTY if it was requested by the session
-	term *terminal
+	// term holds PTY if it was requested by the session.
+	term Terminal
 
-	// agent is a client to remote SSH agent
+	// agent is a client to remote SSH agent.
 	agent agent.Agent
 
-	// agentCh is SSH channel using SSH agent protocol
-	agentCh ssh.Channel
+	// agentCh is SSH channel using SSH agent protocol.
+	agentChannel ssh.Channel
```

Comment:
`agentCh` is more consistent with what we use

Reply:
Yeah, I've tried to distinguish between a Go channel where we use the `Ch` suffix and a SSH channel. Otherwise it's confusing when you look at variable name.

Reply:
it will be clear from the type though